File: LC 22.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> ans;
        vector<vector<string>> dp(n + 1);
        dp[0] = {""};
        dp[1] = {"()"};
        for (int i = 2; i <= n; ++i) {
            for (int j = 0; j < i; ++j) {
                for (auto& p : dp[j]) {
                    for (auto& q : dp[i-j-1]) {
                        string s = "(" + p + ")" + q;
                        dp[i].emplace_back(s);
                    }
                }
            }
        }
        return dp[n];
    }
};

File: 1000/LC 22.go
func generateParenthesis(n int) (ans []string) {
    var s string
    
    var dfs func(int, int)
    dfs = func(a, b int) {
        if a == 0 && b == 0 {
            ans = append(ans, s)
            return
        }
        if a == b {
            s += "("
            dfs(a - 1, b)
            s = s[:len(s)-1]
        } else {
            if a > 0 {
                s += "("
                dfs(a - 1, b)
                s = s[:len(s)-1]
            }
            s += ")"
            dfs(a, b - 1)
            s = s[:len(s)-1]
        }
    }
    dfs(n, n)

    return 
}
